qir/qat/Rules/Notation/Notation.cpp

```diff
@@ -22,13 +22,48 @@ namespace quantum
         {
             return [](ReplacementRule::Builder&, ReplacementRule::Value* val, ReplacementRule::Captures&,
                       ReplacementRule::Replacements& replacements) {
-                auto type = val->getType();
+                auto type  = val->getType();
+                auto instr = llvm::dyn_cast<llvm::Instruction>(val);
+
+                if (instr == nullptr)
+                {
+                    return false;
+                }
+
                 val->replaceAllUsesWith(llvm::UndefValue::get(type));
-                replacements.push_back({llvm::dyn_cast<llvm::Instruction>(val), nullptr});
+                replacements.push_back({instr, nullptr});
+
                 return true;
             };
         }
 
+        ReplacerFunction deleteUnusedInstruction()
+        {
+            return [](ReplacementRule::Builder&, ReplacementRule::Value* val, ReplacementRule::Captures&,
+                      ReplacementRule::Replacements& replacements) {
+                std::unordered_map<llvm::Value*, llvm::Value*> replace_set{replacements.begin(), replacements.end()};
+
+                auto instr = llvm::dyn_cast<llvm::Instruction>(val);
+
+                if (instr)
+                {
+                    bool is_used = false;
+                    for (auto u : instr->users())
+                    {
+                        is_used |= replace_set.find(u) == replace_set.end();
```

Comment:
`use_empty` is not sufficient because we have a set of instructions which are scheduled for deletion but not deleted yet. Our definition is unused is in terms of the resulting IR rather than the working version which means that while a instruction may well be used in the current working version of the IR, it will be an unused instruction in a few thousand cycles.

Reply:
This sounds a little speculative; I am a bit worried that this introduces either assumptions/dependencies within the code base, or might be more fragile than it needs to. I'll leave it up to you to be the judge here, just weight the benefits of doing it in one pass against potential downsides in code maintainability and robustness.

Reply:
I am happy to have an offline discussion explaining the details of this.